In [2]:
import os
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
visa_df = pd.read_csv('Visadataset.csv')

In [4]:
cat_columns = visa_df.select_dtypes(include='object').columns
num_columns = visa_df.select_dtypes(exclude='object').columns
num_columns

Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object')

In [5]:
empty_data = pd.DataFrame()
for i in num_columns:
    count = len (visa_df[i])
    min_visa=min(visa_df[i])
    max_visa=max(visa_df[i])
    mean_visa=round(visa_df[i].mean(),2)
    median_visa=round(visa_df[i].median(),2)
    std_visa=round(visa_df['prevailing_wage'].std(),2)
    visa_25p=round(np.percentile(visa_df[i],25),2)
    visa_50p=round(np.percentile(visa_df[i],50),2)
    visa_75p=round(np.percentile(visa_df[i],75),2)
    values=[count,min_visa,max_visa,mean_visa,median_visa,std_visa,visa_25p,visa_50p,visa_75p]
    index=['count','min','max','mean','median','std','25p','50p','75p']
    cols=[i]
    new_data = pd.DataFrame(values,index=index,columns=cols)
    empty_data = pd.concat([empty_data,new_data],axis=1)

In [6]:
empty_data

no_of_employees  yr_of_estab  prevailing_wage
count          25480.00     25480.00       25480.0000
min              -26.00      1800.00           2.1367
max           602069.00      2016.00      319210.2700
mean            5667.04      1979.41       74455.8100
median          2109.00      1997.00       70308.2100
std            52815.94     52815.94       52815.9400
25p             1022.00      1976.00       34015.4800
50p             2109.00      1997.00       70308.2100
75p             3504.00      2005.00      107735.5100

In [7]:
np.percentile(visa_df['prevailing_wage'],25)

34015.479999999996

In [8]:
25*25480/100

6370.0

**Percentile**
- 25p:
    - Assume that 25p value = x
    - 25p of data less than value :x
- for eg: prevailing wages of 25p value is 34015.47
- 25p of people having salary less than 34015.47
- total memebers: 25480
- 25p of 25480 = 6370
- which means 6370 peoplo has salary less than 34015.47

- Step 1: Read the prevailing wage
- Step 2: Provide the condition
    - prevailing wage - and its value

In [11]:
wage_data = visa_df['prevailing_wage']
p_25p=np.percentile(visa_df['prevailing_wage'],25)
con = wage_data<p_25p
visa_df[con]

case_id      continent education_of_employee has_job_experience  \
0         EZYV01           Asia           High School                  N   
7         EZYV08  North America            Bachelor's                  Y   
12        EZYV13           Asia            Bachelor's                  Y   
16        EZYV17         Europe              Master's                  Y   
17        EZYV18           Asia              Master's                  Y   
...          ...            ...                   ...                ...   
25461  EZYV25462           Asia              Master's                  Y   
25465  EZYV25466  North America           High School                  N   
25466  EZYV25467         Europe            Bachelor's                  Y   
25470  EZYV25471  North America              Master's                  Y   
25473  EZYV25474           Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
7                         N             3035         1924   
12                        N           123876         1963   
16                        N            76638         1991   
17                        N             2747         2001   
...                     ...              ...          ...   
25461                     N             2861         2004   
25465                     N             2577         1995   
25466                     N             1938         2005   
25470                     N             2272         1970   
25473                     N             2421         2007   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
7                     West         418.2298         Hour                  Y   
12               Northeast       28663.0500         Year                  Y   
16                 Midwest        3706.7900         Year                  Y   
17                    West       16132.6100         Year                  Y   
...                    ...              ...          ...                ...   
25461                 West          54.9196         Hour                  Y   
25465                South         481.2238         Hour                  Y   
25466                 West        6973.4000         Year                  Y   
25470            Northeast         516.4101         Hour                  Y   
25473            Northeast       22845.5600         Year                  Y   

      case_status  
0          Denied  
7          Denied  
12      Certified  
16      Certified  
17      Certified  
...           ...  
25461      Denied  
25465   Certified  
25466      Denied  
25470   Certified  
25473   Certified  

[6370 rows x 12 columns]

In [12]:
50*25480/100

12740.0

In [13]:
p_50p = np.percentile(visa_df['prevailing_wage'],50)
con_50 = wage_data<p_50p
visa_df[con_50]

case_id      continent education_of_employee has_job_experience  \
0         EZYV01           Asia           High School                  N   
6         EZYV07           Asia            Bachelor's                  N   
7         EZYV08  North America            Bachelor's                  Y   
9         EZYV10         Europe             Doctorate                  Y   
12        EZYV13           Asia            Bachelor's                  Y   
...          ...            ...                   ...                ...   
25465  EZYV25466  North America           High School                  N   
25466  EZYV25467         Europe            Bachelor's                  Y   
25470  EZYV25471  North America              Master's                  Y   
25473  EZYV25474           Asia            Bachelor's                  Y   
25474  EZYV25475         Africa             Doctorate                  N   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
6                         N             4985         1994   
7                         N             3035         1924   
9                         N             2251         1995   
12                        N           123876         1963   
...                     ...              ...          ...   
25465                     N             2577         1995   
25466                     N             1938         2005   
25470                     N             2272         1970   
25473                     N             2421         2007   
25474                     N             2594         1979   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
6                    South       53635.3900         Year                  Y   
7                     West         418.2298         Hour                  Y   
9                    South       67514.7600         Year                  Y   
12               Northeast       28663.0500         Year                  Y   
...                    ...              ...          ...                ...   
25465                South         481.2238         Hour                  Y   
25466                 West        6973.4000         Year                  Y   
25470            Northeast         516.4101         Hour                  Y   
25473            Northeast       22845.5600         Year                  Y   
25474            Northeast       51104.7800         Year                  Y   

      case_status  
0          Denied  
6       Certified  
7          Denied  
9       Certified  
12      Certified  
...           ...  
25465   Certified  
25466      Denied  
25470   Certified  
25473   Certified  
25474   Certified  

[12740 rows x 12 columns]

In [14]:
75*25480/100

19110.0

In [15]:
p_75p = np.percentile(visa_df['prevailing_wage'],75)
con_75 = wage_data<p_75p
visa_df[con_75]

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
3         EZYV04      Asia            Bachelor's                  N   
5         EZYV06      Asia              Master's                  Y   
6         EZYV07      Asia            Bachelor's                  N   
...          ...       ...                   ...                ...   
25473  EZYV25474      Asia            Bachelor's                  Y   
25474  EZYV25475    Africa             Doctorate                  N   
25475  EZYV25476      Asia            Bachelor's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
3                         N               98         1897   
5                         N             2339         2012   
6                         N             4985         1994   
...                     ...              ...          ...   
25473                     N             2421         2007   
25474                     N             2594         1979   
25475                     Y             2601         2008   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
3                     West       83434.0300         Year                  Y   
5                    South       78252.1400         Year                  Y   
6                    South       53635.3900         Year                  Y   
...                    ...              ...          ...                ...   
25473            Northeast       22845.5600         Year                  Y   
25474            Northeast       51104.7800         Year                  Y   
25475                South       77092.5700         Year                  Y   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
3          Denied  
5       Certified  
6       Certified  
...           ...  
25473   Certified  
25474   Certified  
25475   Certified  
25478   Certified  
25479   Certified  

[19110 rows x 12 columns]

**Imperical rule**
- (mean - 1std) and (mean + 1std) = 68%
- (mean - 2std) and (mean + 2std) = 95%
- (mean - 3std) and (mean + 3std) = 99.7%

- step 1: calculate the mean of prevailing wage 
- step 2: calculate the std of prevailing wage
- step 3: lb=mean-1*std
- step 4: ub=mean+1*std
- step 5: con1= data<ub
- step 6: con2 = data>lb
- step 7: con3 = con1 and con2
- step 8: data[con3]
- step 9: compare with 68% 25480

In [51]:
mean_wage = visa_df['prevailing_wage'].mean()
sigma_wage = visa_df['prevailing_wage'].std()
low_b_sigma_1 = mean_wage - sigma_wage
upper_b_sigma_1 = mean_wage + sigma_wage
con_1_sigma_1 = visa_df['prevailing_wage']<upper_b_sigma_1
con_2_sigma_1 = visa_df['prevailing_wage']>low_b_sigma_1
con_sigma_1 = con_1_sigma_1 & con_2_sigma_1
len_of_68_per =len(visa_df[con_sigma_1])
# data_68_per = visa_df(con_sigma_1)
print (len_of_68_per)
print (68*visa_df/100)

17171


TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [35]:
low_b_sigma_2 = mean_wage - (2*sigma_wage)
upper_b_sigma_2 = mean_wage + (2*sigma_wage)
con_1_sigma_2 = visa_df['prevailing_wage']<upper_b_sigma_2
con_2_sigma_2 = visa_df['prevailing_wage']>low_b_sigma_2
con_sigma_2 = con_1_sigma_2 & con_2_sigma_2
visa_df[con_sigma_2]

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25474  EZYV25475    Africa             Doctorate                  N   
25475  EZYV25476      Asia            Bachelor's                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25474                     N             2594         1979   
25475                     Y             2601         2008   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25474            Northeast       51104.7800         Year                  Y   
25475                South       77092.5700         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25474   Certified  
25475   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[24582 rows x 12 columns]

In [37]:
low_b_sigma_3 = mean_wage - (3*sigma_wage)
upper_b_sigma_3 = mean_wage + (3*sigma_wage)
con_1_sigma_3 = visa_df['prevailing_wage']<upper_b_sigma_3
con_2_sigma_3 = visa_df['prevailing_wage']>low_b_sigma_3
con_sigma_3 = con_1_sigma_3 & con_2_sigma_3
visa_df[con_sigma_3]

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25474  EZYV25475    Africa             Doctorate                  N   
25475  EZYV25476      Asia            Bachelor's                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25474                     N             2594         1979   
25475                     Y             2601         2008   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25474            Northeast       51104.7800         Year                  Y   
25475                South       77092.5700         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25474   Certified  
25475   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25186 rows x 12 columns]